src/java.base/share/classes/java/math/BigInteger.java

```diff
@@ -2701,6 +2694,130 @@ public BigInteger pow(int exponent) {
         }
     }
 
+    /**
+     * Computes {@code x^n} using repeated squaring trick.
+     * Assumes {@code x != 0}.
+     */
+    private static BigInteger unsignedIntPow(int x, int n) {
+        if (x == 1)
+            return ONE;
+        // Double.PRECISION / bitLength(x) is the largest integer e
+        // such that x^e fits into a double. If e <= 2, we won't use fp arithmetic.
+        // This allows to use fp arithmetic where possible.
+        final int maxExp = Math.max(2, Double.PRECISION / bitLengthForInt(x));
```

Comment:
We should link this 2 to the logic to xToExpAdj below - with comments on both sites.